### ISSUE TITLE
XRunner tests start failing due to timeout

### ISSUE THREAD

**Issue by greenEkatherine** [dotnet/xharness#663]
Since June 24 we noticed many failures for JIT* tests:
![MicrosoftTeams-image (1)](https://user-images.githubusercontent.com/19972657/123806342-208fa400-d8ef-11eb-8377-562deb38b73a.png)

Example build https://dev.azure.com/dnceng/public/_build/results?buildId=1208782&view=logs&j=1fbc174f-133f-57a8-dcec-0de15ad033b4&t=a697e205-a0a3-5096-6a0a-07ceb96d569b

It might be related to XHarness or Arcade update between https://dev.azure.com/dnceng/public/_build/results?buildId=1203823&view=results and https://dev.azure.com/dnceng/public/_build/results?buildId=1203693&view=results

**Comment by greenEkatherine** [comment]
@fanyang-mono please confirm if it's still reproduced

**Comment by fanyang-mono** [comment]
Sadly it is still reproducible for running tests on Android devices. It happened for running tests for both arm64 and arm32.
Here are the builds:
Arm64: https://dev.azure.com/dnceng/public/_build/results?buildId=1237549&view=logs&jobId=1fbc174f-133f-57a8-dcec-0de15ad033b4&j=1fbc174f-133f-57a8-dcec-0de15ad033b4&t=a697e205-a0a3-5096-6a0a-07ceb96d569b
Arm32: https://dev.azure.com/dnceng/public/_build/results?buildId=1236054&view=logs&jobId=1ea9ed70-df96-5876-49f1-ea0dde598d57&j=1ea9ed70-df96-5876-49f1-ea0dde598d57&t=d0d2d285-8a8b-5c9f-4931-7a5bf4a769b2

**Comment by greenEkatherine** [comment]
For both cases it shows space issue. Tests that run later provide empty output so to identify this we need to check the earliest tests in one run. I'm going to check the state of the .devices

**Comment by greenEkatherine** [comment]
I see that there are many JIT* tests left on devices, I'm going to clean it.
@fanyang-mono did you update those PRs to use the latest fix for proper uninstallation?

**Comment by greenEkatherine** [comment]
All devices were cleaned up

Cleanup powershell script for future reference:
```
$packages=& "C:\adb\platform-tools\adb.exe" shell pm list packages net.dot
$parts=$packages.split(':')
For ($i=1; $i -lt $parts.Length; $i+=2) {
    echo $parts[$i] | %{&"C:\adb\platform-tools\adb.exe" uninstall $_}
    }
exit 0
```
@fanyang-mono you may try again but first please make sure that apps properly uninstalled

**Comment by greenEkatherine** [comment]
From chat with @fanyang-mono:

> it is still timing out https://dev.azure.com/dnceng/public/_build/results?buildId=1263712&view=logs&j=1fbc174f-133f-57a8-dcec-0de15ad033b4&t=a697e205-a0a3-5096-6a0a-07ceb96d569b

Now I see that at first we need to set timeout for XHarness less then it is for pipeline job. Otherwise job is killed and we are not able finalize test properly - app left uninstalled and logs weren't copied. Logs will help to investigate an issue further.

Also, I tried to run one of those test APK locally - no timeout, finished with instrumentation failure.

**Comment by premun** [comment]
@greenEkatherine I believe the right way to achieve the timeout buffer would be to make runtime tests use `CustomCommands` and then placing a timeout around this call:
https://github.com/dotnet/arcade/blob/dde80826e2fe464ef34182d2f786a6a51121d9a6/src/Microsoft.DotNet.Helix/Sdk/tools/xharness-runner/xharness-helix-job.android.sh#L75

Then, we would be in control on the SDK side to do cleanup (and later send telemetry) as we could inflate the actual Helix timeout and constrain the injected commands with the timeout provided by the client.

However, I believe runtime team should also constrain on their side too to allow for a proper uninstall on their part.

**Comment by premun** [comment]
Over all, the next steps for the runtime team is to start using the `CustomCommands` and the Helix SDK as intended to give us more control over device runs. This will also be needed for https://github.com/dotnet/core-eng/issues/13607.

/cc @fanyang-mono @SamMonoRT

**Comment by greenEkatherine** [comment]
Thanks to @fanyang-mono that she pointed me out - there is still issue with space. For example https://helix.dot.net/api/jobs/e1d64e1a-4efe-40fb-9ecc-868ff1dbc053/workitems/baseservices.mono?api-version=2019-06-17  attaching output here as well
[baseservices_mono_install.output (1) (1).txt](https://github.com/dotnet/xharness/files/6901535/baseservices_mono_install.output.1.1.txt)

Error `INSTALL_FAILED_INSUFFICIENT_STORAGE: Scanning Failed.: Package net.dot.baseservices_mono could not be assigned a valid UID` means that there are no space to add new user and it seems to be true - `/product` folder in internal storage filled up to 100% on `DNCENGWIN-033`.

However, there are many tests that were running and passed last week on the same machine `DNCENGWIN-033` when our JIT* and baseservices* tests failed there. To confirm it even more I send a test job via JobSender with JIT_jit64.apk to `DNCENGWIN-078` where one of tests also failed. And my installation passed https://helix.dot.net/api/jobs/dc723781-fe5f-4535-928e-f77ce8faca50/workitems/Using%20a%20Payload?api-version=2019-06-17

There must be some difference between these tests or maybe the way they use Helix SDK.

**Comment by greenEkatherine** [comment]
This is related to changes in XHarness (some time before June 24): I just reproduced it with e2e script, then checked out to randomly old version and installation passed

**Comment by greenEkatherine** [comment]
This commit https://github.com/dotnet/xharness/commit/fed4184ebd3315ca717339b2f751693ef8dabd36

**Comment by premun** [comment]
This change only changes `True`/`False` to `true`/`false` if I remember correctly.

I suspect the larger change might be the problem? https://github.com/dotnet/xharness/commit/35f5981ac7ded7c1c95258c0fb86362ed78704b8

Or why does the `ToString` matter here?

**Comment by greenEkatherine** [comment]
> I suspect the larger change might be the problem? [35f5981](https://github.com/dotnet/xharness/commit/35f5981ac7ded7c1c95258c0fb86362ed78704b8)
> 
> Or why does the `ToString` matter here?

Only one commit and this string - I commented it out and installation passed

**Comment by fanyang-mono** [comment]
I wonder if the root cause of this problem is also the root cause of the timeout that we are currently encountering for runtime tests running on iOS. This is one of the work items https://helix.dot.net/api/jobs/ca298f42-4111-4e42-9aaa-57f9e8a90c03/workitems/baseservices.mono?api-version=2019-06-17
The failure looks like this
```
[09:21:51.2101690] Using Xcode 11.5 found in /Applications/Xcode115.app
[09:21:51.2173610] Xamarin.Hosting: Xamarin.Hosting
[09:21:51.2174400] Xamarin.Hosting:     Version: 9acbbed1f6 (main)
[09:21:51.2175040] Xamarin.Hosting:     Xcode: /Applications/Xcode115.app
[09:21:51.2220260] Xamarin.Hosting:     Xcode Version: 11.5
[09:21:51.2230870] Xamarin.Hosting:     Verbosity: 5
[09:21:51.2419570] Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DVTFoundation.framework/Versions/A/DVTFoundation
[09:21:51.2496530] Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DVTiPhoneSimulatorRemoteClient.framework/DVTiPhoneSimulatorRemoteClient
[09:21:51.2498430] Xamarin.Hosting: Loaded /Library/Developer/PrivateFrameworks/CoreSimulator.framework/CoreSimulator
[09:21:51.2500420] Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DTDeviceKitBase.framework/DTDeviceKitBase
[09:21:51.2820120] Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DVTKit.framework/DVTKit
[09:21:51.2829340] Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DTDeviceKit.framework/DTDeviceKit
[09:21:51.2831030] Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DTXConnectionServices.framework/DTXConnectionServices
[09:21:51.2847910] Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DVTSourceControl.framework/DVTSourceControl
[09:21:51.2849880] Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DVTServices.framework/DVTServices
[09:21:51.2852350] Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DVTPortal.framework/DVTPortal
[09:21:51.3396530] Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DVTDocumentation.framework/DVTDocumentation
[09:21:51.3409000] Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DVTAnalyticsClient.framework/DVTAnalyticsClient
[09:21:51.3505740] Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/ContentDeliveryServices.framework/ContentDeliveryServices
[09:21:51.3525250] Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/DVTAnalytics.framework/DVTAnalytics
[09:21:51.3561800] Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/SourceKit.framework/SourceKit
[09:21:51.3814430] Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/Frameworks/IDEFoundation.framework/IDEFoundation
[09:21:51.3831770] Xamarin.Hosting: Loaded /Applications/Xcode115.app/Contents/SharedFrameworks/IDEProducts.framework/IDEProducts
[09:21:51.5359910] Xamarin.Hosting: Simulator watchdogs are not disabled for 'iOS 13.5 (17F61) - iPhone X'.
[09:21:51.5415630] Xamarin.Hosting: Successfully disabled simulator watchdogs for 'iOS 13.5 (17F61) - iPhone X'.
[09:21:51.5615920] Xamarin.Hosting: Launching simulator application 'com.apple.iphonesimulator'
[09:36:49.5210230] Killing process 9921 as it was cancelled
[09:36:49.5260190] Process mlaunch exited with 137
```

**Comment by premun** [comment]
@fanyang-mono as discussed last week - I believe the problem is that the app is not exiting and XHarness waits for it until the timeout comes and it kills the app. I tried running the app manually on the phone and I would expect it to also quit. I saw this for an hour:

![image](https://user-images.githubusercontent.com/7013027/127846619-7f0a90e8-0759-4ce6-bf27-e7e3a48b05b8.png)

**Comment by premun** [comment]
This screenshot is from a run where I used the same arguments as your CI runs ^^

**Comment by fanyang-mono** [comment]
But the run from yesterday on your machine passed. Then I don't know what you were missing from last week's run.

**Comment by SamMonoRT** [comment]
@premun @greenEkatherine  @fanyang-mono  -- what is the status of this  ? seems like we are still blocked by this to merge https://github.com/dotnet/runtime/pull/51642

**Comment by greenEkatherine** [comment]
The root cause for this issue is still unclear for me. Using custom command will allow to set up timeout and get more data for investigation so I'm going to check it again now.
Also, I found an issue with external storage (https://github.com/dotnet/core-eng/issues/13359#issuecomment-924041941) - it wasn't set up and might affect these tests since they require quite weighty space for installation. I will merge the fix soon and see if it helps.

**Comment by SamMonoRT** [comment]
Thanks @greenEkatherine  - please let @fanyang-mono know if you need anything else from her to narrow down the issue.

**Comment by greenEkatherine** [comment]
If I remember correctly, `--package-name` is required parameter, but I see it empty in last failed tests https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-8385bb4e32eb4d64b8/JIT.Math/1/console.cb0d8412.log?sv=2019-07-07&se=2021-10-17T08%3A18%3A52Z&sr=c&sp=rl&sig=sn7vE%2FnIhy2GiUfq1WjXms3zr35s%2BmZlNn2Uk6u9NWw%3D
Might it be the cause of failure?

**Comment by fanyang-mono** [comment]
> If I remember correctly, `--package-name` is required parameter, but I see it empty in last failed tests https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-heads-main-8385bb4e32eb4d64b8/JIT.Math/1/console.cb0d8412.log?sv=2019-07-07&se=2021-10-17T08%3A18%3A52Z&sr=c&sp=rl&sig=sn7vE%2FnIhy2GiUfq1WjXms3zr35s%2BmZlNn2Uk6u9NWw%3D Might it be the cause of failure?

@premun has changed it to optional. https://github.com/dotnet/arcade/pull/7881

**Comment by premun** [comment]
@fanyang-mono I am closing this issue as it no longer reflects any current issue we see in XHarness.

As described here - https://github.com/dotnet/runtime/pull/59772#issuecomment-937933128 - the only timeouts happening are coming from the Xunit runner not calling XHarness in time (might be long running tests or some other problem), however, there were no installation timeouts ever, just failed installations and test timeouts.

We can open a new issue if we find one.
